src/vma/lwip/tcp_out.c

```diff
@@ -603,7 +612,25 @@ tcp_write(struct tcp_pcb *pcb, const void *arg, u32_t len, u8_t is_dummy)
     }
     LWIP_ASSERT("tcp_write: check that first pbuf can hold the complete seglen",
     		(p->len >= seglen));
-    TCP_DATA_COPY2((char *)p->payload + optlen, (u8_t*)arg + pos, seglen, &chksum, &chksum_swapped);
+    if (apiflags & TCP_WRITE_FILE) {
+       piov[piov_cur_index].iov_base = (void *)((char *)p->payload + optlen);
+       piov[piov_cur_index].iov_len = seglen;
+
+       piov_cur_index++;
+       piov_cur_len += seglen;
+       if ((left <= seglen ) || (piov_cur_index >= piov_max_size) || (piov_cur_len >= piov_max_len)) {
+           int ret = 0;
+           int fd = *(int *)arg;
+           ret = sys_readv(fd, piov, piov_cur_index);
```

Comment:
Can readv() read less bytes than requested?

Reply:
Yes,

> Buffers are processed in array order.  This means that readv() com‐
       pletely fills iov[0] before proceeding to iov[1], and so on.  (If
       there is insufficient data, then not all buffers pointed to by iov
       may be filled.)